master/internal/db/bun.go

```diff
@@ -0,0 +1,115 @@
+package db
+
+import (
+	"database/sql"
+	"context"
+
+	"github.com/uptrace/bun"
+	"github.com/uptrace/bun/dialect/pgdialect"
+	"github.com/uptrace/bun/extra/bundebug"
+
+	// "github.com/determined-ai/determined/master/pkg/model"
+)
+
+// dbLike is either a *bun.DB or a *bun.Tx.  This is what our dbImpl struct depends on to provide
+// the BunDB interface.
+type dbLike interface {
+	NewSelect() *bun.SelectQuery
+	NewInsert() *bun.InsertQuery
+	NewUpdate() *bun.UpdateQuery
+	NewDelete() *bun.DeleteQuery
+}
+
+// dbImpl implements BunDB.  Utilmately, dbImpl allows us to have identical OO-behavior on either
+// *bun.DB objects or bun.Tx objects.
+type dbImpl struct {
+	dbLike
+}
+
+// RootBunDB provides the BunDB interface, in addition to a RunInTx() method.  RunInTx() is not
+// nestable because the funcion arg passed to RunInTx will not recieve a RootBunDB, but a bare
+// BunDB object based on the bun.Tx.
+type RootBunDB struct {
+	// bun.DB makes RootBunDB implement dbLike.
+	*bun.DB
+	// dbImpl makes RootBunDB implement BunDB.  This will always just contain the same *bun.DB.
+	dbImpl
+}
+
+// NewRootBunDB creates a RootBunDB from a postgres sql connection.
+func NewRootBunDB(db *sql.DB) RootBunDB {
+	bunDB := bun.NewDB(db, pgdialect.New())
+	// bunDB.AddQueryHook(bundebug.NewQueryHook(bundebug.WithVerbose(true)))
+	bunDB.AddQueryHook(bundebug.NewQueryHook())
+	return RootBunDB{bunDB, dbImpl{bunDB}}
+}
+
+// RootBunDB wraps bun.DB.RunInTx, but provides a BunDB interface without giving access to the
+// actual bun.Tx object.
+func (db *RootBunDB) RunInTx(ctx context.Context, fn func(context.Context, BunDB) error) error {
+	wrapperFn := func(ctx_ context.Context, tx bun.Tx) error {
+		return fn(ctx_, dbImpl{tx})
+	}
+	return db.DB.RunInTx(ctx, nil, wrapperFn)
+}
+
+// QueryExt is a helper struct to define custom .Where, .Order, and .Limit without a ton of inlined
+// functions.
+type QueryExt struct {
```

Comment:
I am not super excited about this solution. asymptotically it's a custom proxy that'll need to handle the entirety of bun options.

Couple thoughts/options:
1. Functions like `CheckpointsList` that use this doesn't add much before/after the query: basically only `db.NewSelect().Model(...)` call before and error handling after. is that worth the extra complexity? both can be done inline in the "view" methods.
2. We could have `filterFn` approach similar to bun's `RunInTx` interface. i.e. 
signature
```
func CheckpointsList(ctx context.Context, filterFn func(bun.SelectQuery) (bun.SelectQuery, error))
```
and then at call-site
```
# req is grpc request thing
ckpts, err := CheckpointsList(ctx, func(ctx context.Context, query bun.SelectQuery) {
  	query = query.Where("trial_id = ?", req.Id)
    ...
    return query, nil
})

Reply:
I experimented with the `filterFn` but I really disliked it, because there are cases where you hit errors _creating_ the filter (not running the filter), but when when you do `ckpts, err := ckpts.List(ctx, func(*bun.SelectQuery)(*bun.SelectQuery, error))`, you don't know if your `err` came from user error (defining a nonsensical error) or from the database call itself.  This would be the difference between an invalid request error and a 500 error, so the calling code does need to distinguish.

I toyed with nesting closures, but it got pretty unreadable pretty quick.

Reply:
>  This would be the difference between an invalid request error and a 500 error, so the calling code does need to distinguish.

At what point should we be building an invalid query because of user input?

Reply:
If for some reason we are, then we could define the signature as like
```
func CheckpointsList(ctx context.Context, filterFn func(bun.SelectQuery) (bun.SelectQuery, errInvalidQuery))
```
and use `errors.Is`.